core/modules/replica/ChunkLocker.cc

```diff
@@ -0,0 +1,183 @@
+
+    ChunksByOwners owner2chunks;
+    if (_owner2chunks.count(owner)) {
+        owner2chunks[owner] = _owner2chunks.at(owner);
+    }
+    return owner2chunks;
+}
+
+bool ChunkLocker::lock(Chunk const&       chunk,
+                       std::string const& owner) {
+
+    util::Lock mLock(_mtx, "ChunkLocker::lock");
+
+    if (owner.empty()) {
+        throw std::invalid_argument("ChunkLocker::lock  empty owner");
+    }
+    if (_chunk2owner.count(chunk)) {
+        return owner == _chunk2owner.at(chunk);
+    }
+    _chunk2owner [chunk] = owner;
+    _owner2chunks[owner].push_back(chunk);
+
+    return true;
+}
+
+bool ChunkLocker::release(Chunk const& chunk) {
+
+    util::Lock mLock(_mtx, "ChunkLocker::release(chunk)");
+
+    // An owner (if set) will be ignored by the current method
+    std::string owner;
+    return releaseImpl(mLock, chunk, owner);
+}
+
+bool ChunkLocker::release(Chunk const& chunk,
+                          std::string& owner) {
+
+    util::Lock mLock(_mtx, "ChunkLocker::release(chunk,owner)");
+    return releaseImpl(mLock, chunk, owner);
+}
+
+bool ChunkLocker::releaseImpl(util::Lock const& mLock,
+                              Chunk const& chunk,
+                              std::string& owner) {
+
+    if (not _chunk2owner.count(chunk)) return false;
+
+    // Remove the chunk from this map _only_ after getting its owner
+    owner = _chunk2owner.at(chunk);
+    _chunk2owner.erase(chunk);
+
+    // Remove the chunk from the list of all chunks claimed by that particular
+    // owner as well.
+    std::list<Chunk>& chunks = _owner2chunks.at(owner);
+    chunks.remove(chunk);
+
+    // This last step is needed to avoid building up empty lists
+    // of non-existing owners
+    if (chunks.empty()) _owner2chunks.erase(owner);
+
+    return true;
+}
+
+std::vector<Chunk> ChunkLocker::release(std::string const& owner) {
+
+    util::Lock mLock(_mtx, "ChunkLocker::release(owner)");
+
+    if (owner.empty()) {
+        throw std::invalid_argument("ChunkLocker::release  empty owner");
+    }
+
+    // First get all chunks claimed by the specified owner.
+    // This list s also going to be returned o teh cller.
+    std::vector<Chunk> chunks;
+    if (_owner2chunks.count(owner)) {
+        for (auto&& chunk: _owner2chunks.at(owner)) {
+            chunks.push_back(chunk);
+        }
+    }
+
+    // Then release the select chunks
+    for (auto&& chunk: chunks) {
+        release(chunk);
+    }
+    return chunks;
+}
+
+//////////////////////////////////////
+//               misc               //
+//////////////////////////////////////
+
+std::ostream& operator <<(std::ostream& os, ChunkLocker::ChunksByOwners const& chunks) {
+    for (auto&& entry: chunks) {
+        std::string const& owner = entry.first;
+        os  << "Chunk owner: " << owner << "\n";
+        for (Chunk const& chunk: entry.second) {
+            os  << "    " << chunk.databaseFamily << ":" << chunk.number << "\n";
```

Comment:
Chunk has its own operator<<, it would be better to just call that here.

Reply:
Though, your statement was 100% correct :-)